src/System.Runtime.WindowsRuntime/tests/System/IO/AsWinRTStreamTests.cs

```diff
@@ -0,0 +1,396 @@
+
+                // bytesCompleted must be in range [0, maxBytesToRead] asyncReadOp in progress
+                Assert.InRange(bytesCompleted, 0u, (uint)TestStreamProvider.ModelStreamLength);
+            };
+
+            readOp.Completed = (asyncReadOp, passedStatus) =>
+            {
+                try
+                {
+                    completedCallbackInvoked = true;
+
+                    // asyncReadOp.Id in a completion callback must match the ID of the asyncReadOp to which the callback was assigned
+                    Assert.Equal(readOpId, asyncReadOp.Id);
+
+                    // asyncReadOp.Status must match passedStatus for a completed asyncReadOp
+                    Assert.Equal(passedStatus, asyncReadOp.Status);
+
+                    // asyncReadOp.Status must be 'Completed' for a completed asyncReadOp
+                    Assert.Equal(AsyncStatus.Completed, asyncReadOp.Status);
+
+                    IBuffer resultBuffer = asyncReadOp.GetResults();
+
+                    // asyncReadOp.GetResults() must not return null for a completed asyncReadOp
+                    Assert.NotNull(resultBuffer);
+
+                    AssertExtensions.GreaterThan(resultBuffer.Capacity, 0u, "resultBuffer.Capacity should be more than zero in completed callback");
+                    AssertExtensions.GreaterThan(resultBuffer.Length, 0u, "resultBuffer.Length should be more than zero in completed callback");
+                    AssertExtensions.LessThanOrEqualTo(resultBuffer.Length, resultBuffer.Capacity, "resultBuffer.Length should be <= Capacity in completed callback");
+
+                    if (inputStreamOptions == InputStreamOptions.None)
+                    {
+                        // resultBuffer.Length must be equal to requested number of bytes when an asyncReadOp with
+                        // InputStreamOptions.None completes successfully
+                        Assert.Equal(resultBuffer.Length, (uint)TestStreamProvider.ModelStreamLength);
+                    }
+
+                    if (inputStreamOptions == InputStreamOptions.Partial)
+                    {
+                        AssertExtensions.LessThanOrEqualTo(resultBuffer.Length, (uint)TestStreamProvider.ModelStreamLength,
+                            "resultBuffer.Length must be <= requested number of bytes with InputStreamOptions.Partial in completed callback");
+                    }
+                    buffer = resultBuffer;
+                }
+                finally
+                {
+                    waitHandle.Set();
+                }
+            };
+
+            // Now, let's block until the read op is complete.
+            // We speculate that it will complete within 3500 msec, although under high load it may not be.
+            // If the test fails we should use a better way to determine if callback is really not invoked, or if it's just too slow.
+            waitHandle.WaitOne(500);
```

Comment:
What is the purpose of multiple WaitOne? The value is a timeout, not a wait interval. You could just do `waitHandle.WaitOne(3500)`